extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,270 @@
+    }
+
+    public void write(int b) throws IOException {
+        write(new byte[] {(byte) b}, 0, 1);
+    }
+
+    public void write(byte[] b) throws IOException {
+        write(b, 0, b.length);
+    }
+
+    public void write(final byte @NotNull [] b, int off, int len) throws IOException {
+        while (len != 0) {
+            if (uploadId == null) {
+                // Initialize the upload ID for the multipart upload
+                uploadId = initiateMultipartUpload();
+            }
+
+            // We use buffers and futures in a round-robin fashion
+            final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
+            if (pendingRequests.size() == nextSlotId) {
+                pendingRequests.add(new OutgoingRequest(partSize));
+            } else if (pendingRequests.size() < nextSlotId) {
+                throw new IllegalStateException("Unexpected slot ID " + nextSlotId + " for uri " + uri + " with " +
+                        pendingRequests.size() + " pending requests.");
+            }
+
+            // Wait for the oldest upload to complete if no space is available
+            final OutgoingRequest useRequest = pendingRequests.get(nextSlotId);
+            if (useRequest.future != null) {
+                waitForCompletion(useRequest);
+            }
+
+            // Write as much as possible to this buffer
+            final ByteBuffer buffer = useRequest.buffer;
+            final int remaining = buffer.remaining();
+            if (remaining >= len) {
+                buffer.put(b, off, len);
+                if (!buffer.hasRemaining()) {
+                    sendPartRequest(useRequest);
+                }
+                break; // done
+            }
+            buffer.put(b, off, remaining);
+            sendPartRequest(useRequest);
+            off += remaining;
+            len -= remaining;
+        }
+    }
+
+    public void flush() throws IOException {
+        final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
+        if (pendingRequests.size() == nextSlotId) {
+            // Nothing to flush
+            return;
+        }
+        final OutgoingRequest request = pendingRequests.get(nextSlotId);
+        if (request.buffer.position() != 0 && request.future == null) {
+            sendPartRequest(request);
+        }
+    }
+
+    /**
+     * Try to finish the multipart upload and close the stream. Cancel the upload if an error occurs.
+     *
+     * @throws IOException if an error occurs while closing the stream
+     */
+    public void close() throws IOException {
+        if (uploadId == null) {
+            return;
+        }
+        try {
+            flush();
+            completeMultipartUpload();
+        } catch (final IOException e) {
+            abort();
+            throw new IOException(String.format("Error closing S3OutputStream for uri %s, aborting upload.", uri), e);
+        }
+        uploadId = null;
+    }
```

Comment:
I'm trying to consider this in the context of some sort of failure in the writing stages:

```java
try (final OutputStream out = new S3OutputStream(...)) {
  out.write(...); // throws exception
}
```

I just need to convince myself that we'll always abort correctly...